File: 1320-remove-all-adjacent-duplicates-in-string-ii/1320-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        int n = s.size();
        stack<pair<char,int>> st;

        for(int i = 0; i < n; i++){
            char curr = s[i];

            if(!st.empty())
            {
                if(st.top().first == curr)
                {
                    st.top().second += 1;
                    if(st.top().second % k == 0)
                    {
                        st.pop();
                    }
                }
                else{
                    st.push({curr, 1});
                }
            }
            else{
                st.push({curr, 1});
            }
        }

        string ans = "";
        while(!st.empty())
        {
            ans = string(st.top().second, st.top().first)+ans;
            st.pop();
        }

        return ans;
    }
};

File: 1320-remove-all-adjacent-duplicates-in-string-ii/1320-remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    public String removeDuplicates(String s, int k) {
        StringBuilder sb = new StringBuilder(s);
        Stack<Integer> counts = new Stack<>();
        for (int i = 0; i < sb.length(); ++i) {
            if (i == 0 || sb.charAt(i) != sb.charAt(i - 1)) {
                counts.push(1);
            } else {
                int incremented = counts.pop() + 1;
                if (incremented == k) {
                    sb.delete(i - k + 1, i + 1);
                    i = i - k;
                } else {
                    counts.push(incremented);
                }
            }
        }
        return sb.toString();
    }
}
/*
class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Integer> counts = new Stack<>();
        char[] sa = s.toCharArray();
        int j = 0;
        for (int i = 0; i < s.length(); ++i, ++j) {
            sa[j] = sa[i];
            if (j == 0 || sa[j] != sa[j - 1]) {
                counts.push(1);
            } else {
                int incremented = counts.pop() + 1;
                if (incremented == k) {
                    j = j - k;
                } else {
                    counts.push(incremented);
                }
            }
        }
        return new String(sa, 0, j);
    }
}

*/